### ISSUE TITLE
addcdiv computes incorrect results on MPS with noncontiguous tensors

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#118115]
### 🐛 Describe the bug

I discovered the following when enabling optimizer tests on MPS (https://github.com/pytorch/pytorch/pull/117410) and discovered that all the optimizers with addcdiv were failing

Repro:
```python
import torch

x = torch.rand(2, 3, device="cpu").t()
x_mps = x.to(device="mps")
print(x)
print(x_mps)  # same

y = torch.rand(2, 3, device="cpu").t()
y_mps = y.to(device="mps")
print(y)
print(y_mps)  # same

z = torch.rand_like(x)
z_mps = z.to(device="mps")
print(z)
print(z_mps)  # same

print(x.addcdiv_(y, z))
print(x_mps.addcdiv_(y_mps, z_mps)) # NOT same!
```

### Versions

on main

cc @ezyang @gchanan @zou3519 @kadeng @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

**Comment by lamroger** [comment]
I poked around to see if I could find the issue. 

```
    def test_addcdiv_transpose(self):
        x = torch.randn(2, 3, device="cpu").t()
        x_mps = x.to(device="mps")

        print(x)
        print(x_mps)

        y = torch.ones(3, 2, device="cpu")
        y_mps = y.detach().clone().to(device="mps")

        z = torch.ones(3, 2, device="cpu")
        z_mps = z.detach().clone().to(device="mps")

        print("cpu:", x.addcdiv_(y, z, value=1))
        print("mps inplace:", x_mps.addcdiv_(y_mps, z_mps, value=1))
        print("mps:", torch.addcdiv(x_mps, y_mps, z_mps, value=1))
        self.assertEqual(x.addcdiv_(y, z), x_mps.addcdiv_(y_mps, z_mps))
```

```
test/test_mps.py tensor([[ 0.0461,  0.2167],
        [ 0.4024, -0.6123],
        [-1.0115,  0.5036]])
tensor([[ 0.0461,  0.2167],
        [ 0.4024, -0.6123],
        [-1.0115,  0.5036]], device='mps:0')
cpu: tensor([[ 1.0461,  1.2167],
        [ 1.4024,  0.3877],
        [-0.0115,  1.5036]])
mps inplace: tensor([[ 0.0461,  0.2167],
        [ 0.4024, -0.6123],
        [-1.0115,  0.5036]], device='mps:0')
mps: tensor([[ 0.0461, -0.6123],
        [ 0.2167, -1.0115],
        [ 0.4024,  0.5036]], device='mps:0')
```

Looks like it's an issue with the instance self being transposed. When it's not transposed, it passes.

Down to help fix. Is there a flag that can help trace the graph to find where it's acting up? 
Also new to this so any explanation of how optimizers work would be helpful.

**Comment by lamroger** [comment]
@janeyx99 Hi - what do you think about adding a `TORCH_CHECK(self.is_contiguous(), "self must be contiguous");
` to addc_mul_div_out_mps for now? 

Making `x_mps = x.to(device="mps").contiguous()` gives the expected result. 

Interested in getting to the root cause but I don't know enough about MPS and how it deals with strides.

**Comment by janeyx99** [comment]
Hi @lamroger--thanks for making a try to fix this. @malfet  is the general plan for fixing these to add `.contiguous` for correctness at the expense of perf? Or is there a longer term plan where MPS will performantly handle contiguity?

**Comment by kulinseth** [comment]
@janeyx99 , similar to this #https://github.com/pytorch/pytorch/pull/123049 , we can add a way to insert gather before we materialize the memory in the Tensor to have addcdiv operate on.

**Comment by janeyx99** [comment]
@kulinseth for my understanding, it means that MPS kernels can only handle contiguous memory and so noncontiguity will always result in an extra copy before calling an op in mps?

**Comment by kulinseth** [comment]
> @kulinseth for my understanding, it means that MPS kernels can only handle contiguous memory and so noncontiguity will always result in an extra copy before calling an op in mps?

We don’t make a copy till we need to materialize the memory such as before gemm or conv operations . Currently for handling operations we do make a copy. But we are working on a solution to fix this by passing view tensors directly to the operations which will avoid the copy

**Comment by jhavukainen** [comment]
This issue is fixed with the merging of #124442
